Persist an in-progress game across page loads.

StasJS opened the pull request:
**UPDATED - see discussion below.**

This is a WIP, because there are obviously no new tests, and because there are still some kinks to iron out. I was hoping what I had here was enough to get some early feedback!

Despite what was suggested #23, I've preserved the undo/redo stacks. There's a bit of data being serialised/deserialised as a result, but probably isn't that taxing in the grand scheme of things. I imagine if we had to do something for performance reasons we could add it to a queue to be processed asynchronously by a service worker, or something like that?.

Things to work out:
- There's some hairiness in trying to maintain the timer across page loads etc. I haven't really thought about it too hard, but there are a couple of comments in the code relating to my first (naive) approach which would be to march the `startTime` forward using the fact that when we persist game state to local storage, we generate a timestamp. Because of the timestamp, we kind of have an idea of how long the user was playing before the page unloaded. So we _could_ subtract that timespan from the current time when the user resumes their game, and consider that the new `startTime`. That seems less than ideal though.
- As seen at the very end of the GIF, there's a weird quirk where the `currentSnapshot` and `undoList` de-sync. In the GIF, you can see that I made progress on the sudoku, then exited and resumed the session. The snapshot state looks correct after resuming, but clicking the undo button pushes us back two actions, rather than one. I haven't figured out a set of steps that makes the de-sync totally reproducible, but I can see it after playing around in the sudoku and doing some number of session-resumes. I'm really hoping you have some insight here!
![sudoku-web-app-session-persistence](https://user-images.githubusercontent.com/24515806/125283811-de457880-e36c-11eb-805b-19d1c7d4c67d.gif)

StasJS commented:
I'd appreciate a look @grantm if/when you have some time.

grantm commented:
Good work on bending my code to your will.

- I agree that preserving the undo/redo stacks is valuable and doesn't need to be too big a deal.  I might a bit about encoding a more compact "diff" between snapshots - the list of snapshots has a lot of redundant data.

- I'm now also wondering if it might be worth supporting tracking more than one game in progress.  Perhaps using the initial digits as part of the key to find the saved game state.

- Your approach to handling time by recalculating the start time seems perfectly reasonable. In fact I think that's how I implemented pause/unpause.

- Your change to expose the serialised game state in a data attribute on the body tag has also removed the code that was exposing the currentSnapshot.  Not sure if that was a) deliberate and b) intended to be part of the complete solution, but we need to continue to store the current snapshot in the same format it is currently as that is there for use by bookmarklets. So maybe do both?

- You've removed the 'beforeunload' handler but I think it's definitely worth retaining.  Sure, with the game state saved, accidentally closing a tab won't be the end of the world, but it could still be inconvenient.  I still get saved regularly by the onunload handler.

- When restoring a game, the URL needs to change to include the initial digits.  That might require some gymnastics.  I got away without using a router initially, but may need to add one eventually.

- I actually don't have any brilliant insight into the issue with undo/redo, but I suspect it's the sort of thing that would become clear with some unit tests to exercise the expected state changes.

StasJS commented:
Thanks for that feedback!
 I'll try fill out a bit more of the functionality. I think keeping it to 1 game seems like a reasonable first stab IMO.

>Your change to expose the serialised game state in a data attribute on the body tag has also removed the code that was exposing the currentSnapshot. Not sure if that was a) deliberate and b) intended to be part of the complete solution, but we need to continue to store the current snapshot in the same format it is currently as that is there for use by bookmarklets. So maybe do both?

I see, I did a string search and only saw it used by the beforeunload handler - must have missed something! I'll re-instate that logic :) as well the beforeunload handler.


I'll work on adding the URL and recalculating start time too.

> I actually don't have any brilliant insight into the issue with undo/redo, but I suspect it's the sort of thing that would become clear with some unit tests to exercise the expected state changes.

I hope so too 😆 definitely on the agenda as well though I haven't though about how best to test this stuff 😞 

Might be a week or two before I update the PR - hopefully that's not a problem

grantm commented:
By the way I merged the part of your commit that was updating the `isList()` (etc.) methods in `not-mutable.js`. That was definitely a useful improvement.

StasJS commented:
I've updated the PR with a number of key changes

1. I've reinstated the `beforeunload` handler, and the saving of `currentSnapshot` to the DOM.
2. I've seriously rearranged and refactored the code that saves to and reads gamestate from local storage - it's now more concentrated in `sudoku-model.js` rather than in components.
3. When resuming an existing game, you'll now see the `initialDigits` appearing in the route (as well as a flag to tell the `App` component 'yes, please load a grid from local storage if you can find one'.
4. The timer works as you'd expect across resumes. If you're paused, close the tab, and resume, the timer recalculates elapsed time using the paused time. If you're not paused, close the tab, and resume, the timer recalculates elapsed time using the last gamestate save's timestamp.
5. I fixed the bug I highlighted in my initial post. We weren't using the 'save grid data' callback often enough (e.g. on undo/redos), so we were seeing a mismatch between `currentSnapshot` and other parts of the state. I think an observable pattern would be more scalable, but for the moment the solution is to sprinkle `notifyGamestateChange` (which saves to DOM/local storage) in key methods.

I've played around with the App a whole bunch, and I can't see any obvious issues. The current tests pass also.
I still haven't written any tests yet. I would be kind of keen to get some advise on how far you think I should go in the testing department. To me it feels like there are tests to write around
- saving gamestate
- loading gamestate (timers, undo/redo lists, snapshot)
- `App` handling the new branching logic for grid initialisation.

Can you think of anything else?

StasJS commented on src/components/app/app.js:
```diff
@@ -52,13 +52,27 @@ const inputModeFromHotKey = {
+        }
     });
+
+    if (params.get('r') === "1") {
```
`r` is the flag that tells the app we should attempt to load a grid from local storage.

StasJS commented on src/components/modal/modal-welcome.js:
```diff
@@ -63,6 +63,39 @@ function RecentlyShared({modalState}) {
+
+    const restoreLocalSessionHandler = () => {
+        const level = modelHelpers.difficultyLevelName(serialisedGrid.difficultyLevel);
+        window.location.search = `s=${serialisedGrid.initialDigits}&d=${level}&r=1`
```
this now just pushes a new URL with `r=1` to trigger the new logic

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -498,40 +498,48 @@ export const modelHelpers = {
-            const undoList = grid.get('undoList');
-            if (actionsBlocked(grid) || undoList.size < 1) {
-                return grid;
+        return modelHelpers.retainSelection(grid, (g) => {
```
renamed to `g` to avoid javascript closure confusion

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -498,40 +498,48 @@ export const modelHelpers = {
                 .update('redoList', list => list.push(beforeUndo));
-            return modelHelpers.checkCompletedDigits(grid);
+            g = modelHelpers.checkCompletedDigits(g);
+            modelHelpers.notifyGamestateChange(g);
```
this now properly causes an update to the state in local storage.

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -498,40 +498,48 @@ export const modelHelpers = {
                 .update('undoList', list => list.push(beforeRedo));
-            return modelHelpers.checkCompletedDigits(grid);
+            g = modelHelpers.checkCompletedDigits(g);
+            modelHelpers.notifyGamestateChange(g);
```
same as undo

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -1083,21 +1143,25 @@ export const modelHelpers = {
                 escapeAction: 'resume-timer',
             },
         });
+        modelHelpers.notifyGamestateChange(g);
```
pausing relates to `pausedAt` which we need for resume logic, so call back

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -1083,21 +1143,25 @@ export const modelHelpers = {
             pausedAt: undefined,
             startTime: Date.now() - elapsed,
         });
+        modelHelpers.notifyGamestateChange(g);
```
same here

StasJS commented on src/lib/sudoku-model.js:
```diff
@@ -1135,9 +1199,11 @@ export const modelHelpers = {
+        grid = modelHelpers.applySelectionOp(grid, 'clearSelection')
             .set('solved', true)
             .set('endTime', Date.now());
+        modelHelpers.notifyGamestateChange(grid);
```
same here

grantm commented:
I haven't had a chance to look too closely at the updated state of things but the changes you describe sound pretty solid. I'm not too worried about tests at this stage and if you're otherwise happy with it then feel free to remove the "draft" status. When go through and do the merge, if I spot anything that needs fixing I can just fix it.

StasJS commented:
should be ready to review then :) Did a final cleanup of the timer logic to make sure it tracks accurately between session restores.

StasJS commented:
Hey @grantm is there anything I can do to help the review of this? I understand that this is a pretty sizable PR

grantm commented:
Hi Jordan. Sorry this has taken a while. I have merged your changes in my dev environment and I'm readying a release. The delay is simply because I've had a lot on, so not much spare time for this project. It will happen 😄
